### ISSUE TITLE
Build failed with Clang 6.0

### ISSUE THREAD

**Issue by kasper3** [dotnet/coreclr#16850]
I am getting some errors on ubuntu when building coreclr with clang 6.0.

```sh
[  0%] Building CXX object src/corefx/System.Globalization.Native/CMakeFiles/System.Globalization.Native_Static.dir/__/__/__/version.cpp.o
[  0%] Building CXX object src/corefx/System.Globalization.Native/CMakeFiles/System.Globalization.Native.dir/__/__/__/version.cpp.o
[  0%] Linking CXX static library System.Globalization.Native.a
[  0%] Linking CXX shared library System.Globalization.Native.so
[  0%] Built target System.Globalization.Native_Static
Scanning dependencies of target coreclrpal
Verifying System.Globalization.Native dependencies
[  0%] Built target System.Globalization.Native
Scanning dependencies of target eventprovider
[  0%] Building CXX object src/pal/src/eventprovider/CMakeFiles/eventprovider.dir/lttng/eventprovdotnetruntime.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/file.cpp.o
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/eventprovdotnetruntime.cpp:24:
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/tpdotnetruntime.h:14:
In file included from /coreclr/src/pal/inc/rt/palrt.h:1617:
/coreclr/src/pal/inc/rt/ntimage.h:47:10: error: the current #pragma pack aligment value is modified in the included file
      [-Werror,-Wpragma-pack]
#include "pshpack4.h"                   // 4 byte packing is the default
         ^
/coreclr/src/pal/inc/rt/pshpack4.h:36:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack(4)
        ^
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/eventprovdotnetruntime.cpp:24:
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/tpdotnetruntime.h:14:
In file included from /coreclr/src/pal/inc/rt/palrt.h:1617:
/coreclr/src/pal/inc/rt/ntimage.h:55:10: error: the current #pragma pack aligment value is modified in the included file
      [-Werror,-Wpragma-pack]
#include "pshpack2.h"                   // 16 bit headers are 2 byte packed
         ^
/coreclr/src/pal/inc/rt/pshpack2.h:35:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack(2)
        ^
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/eventprovdotnetruntime.cpp:24:
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/tpdotnetruntime.h:14:
In file included from /coreclr/src/pal/inc/rt/palrt.h:1617:
/coreclr/src/pal/inc/rt/ntimage.h:177:10: error: the current #pragma pack aligment value is modified in the included
      file [-Werror,-Wpragma-pack]
#include "poppack.h"                    // Back to 4 byte packing
         ^
/coreclr/src/pal/inc/rt/poppack.h:37:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack()
        ^
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/eventprovdotnetruntime.cpp:24:
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/tpdotnetruntime.h:14:
In file included from /coreclr/src/pal/inc/rt/palrt.h:1617:
/coreclr/src/pal/inc/rt/ntimage.h:571:10: error: the current #pragma pack aligment value is modified in the included
      file [-Werror,-Wpragma-pack]
#include "pshpack2.h"                       // Symbols, relocs, and linenumbers are 2 byte packed
         ^
/coreclr/src/pal/inc/rt/pshpack2.h:35:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack(2)
        ^
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/eventprovdotnetruntime.cpp:24:
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/tpdotnetruntime.h:14:
In file included from /coreclr/src/pal/inc/rt/palrt.h:1617:
/coreclr/src/pal/inc/rt/ntimage.h:773:10: error: the current #pragma pack aligment value is modified in the included
      file [-Werror,-Wpragma-pack]
#include "poppack.h"
         ^
/coreclr/src/pal/inc/rt/poppack.h:37:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack()
        ^
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/eventprovdotnetruntime.cpp:24:
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/tpdotnetruntime.h:14:
In file included from /coreclr/src/pal/inc/rt/palrt.h:1617:
/coreclr/src/pal/inc/rt/ntimage.h:1227:10: error: the current #pragma pack aligment value is modified in the included
      file [-Werror,-Wpragma-pack]
#include "pshpack8.h"                       // Use align 8 for the 64-bit IAT.
         ^
/coreclr/src/pal/inc/rt/pshpack8.h:36:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack(8)
        ^
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/eventprovdotnetruntime.cpp:24:
In file included from /coreclr/bin/obj/Linux.x64.Debug/eventing/eventprovider/lttng/tpdotnetruntime.h:14:
In file included from /coreclr/src/pal/inc/rt/palrt.h:1617:
/coreclr/src/pal/inc/rt/ntimage.h:1239:10: error: the current #pragma pack aligment value is modified in the included
      file [-Werror,-Wpragma-pack]
#include "poppack.h"                        // Back to 4 byte packing
         ^
/coreclr/src/pal/inc/rt/poppack.h:37:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack()
        ^
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/filecrt.cpp.o
7 errors generated.
src/pal/src/eventprovider/CMakeFiles/eventprovider.dir/build.make:62: recipe for target 'src/pal/src/eventprovider/CMakeFiles/eventprovider.dir/lttng/eventprovdotnetruntime.cpp.o' failed
make[2]: *** [src/pal/src/eventprovider/CMakeFiles/eventprovider.dir/lttng/eventprovdotnetruntime.cpp.o] Error 1
CMakeFiles/Makefile2:353: recipe for target 'src/pal/src/eventprovider/CMakeFiles/eventprovider.dir/all' failed
make[1]: *** [src/pal/src/eventprovider/CMakeFiles/eventprovider.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/lstr.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/malloc.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/math.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/mbstring.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/misc.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/misctls.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/path.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/printf.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/printfcpp.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/silent_printf.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/string.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/stringtls.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/thread.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/wchar.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/cruntime/wchartls.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/debug/debug.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/exception/seh.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/exception/signal.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/exception/remote-unwind.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/file/directory.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/file/file.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/file/filetime.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/file/find.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/file/path.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/handlemgr/handleapi.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/handlemgr/handlemgr.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/init/pal.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/init/sxs.cpp.o
[  0%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/loader/module.cpp.o
[  1%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/loader/modulename.cpp.o
[  1%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/locale/unicode.cpp.o
[  1%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/locale/unicode_data.cpp.o
[  1%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/locale/utf8.cpp.o
[  1%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/map/common.cpp.o
[  1%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/map/map.cpp.o
[  1%] Building CXX object src/pal/src/CMakeFiles/coreclrpal.dir/map/virtual.cpp.o
In file included from /coreclr/src/pal/src/map/map.cpp:40:
/coreclr/src/pal/inc/rt/ntimage.h:47:10: error: the current #pragma pack aligment value is modified in the included file
      [-Werror,-Wpragma-pack]
#include "pshpack4.h"                   // 4 byte packing is the default
         ^
/coreclr/src/pal/inc/rt/pshpack4.h:36:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack(4)
        ^
In file included from /coreclr/src/pal/src/map/map.cpp:40:
/coreclr/src/pal/inc/rt/ntimage.h:55:10: error: the current #pragma pack aligment value is modified in the included file
      [-Werror,-Wpragma-pack]
#include "pshpack2.h"                   // 16 bit headers are 2 byte packed
         ^
/coreclr/src/pal/inc/rt/pshpack2.h:35:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack(2)
        ^
In file included from /coreclr/src/pal/src/map/map.cpp:40:
/coreclr/src/pal/inc/rt/ntimage.h:177:10: error: the current #pragma pack aligment value is modified in the included
      file [-Werror,-Wpragma-pack]
#include "poppack.h"                    // Back to 4 byte packing
         ^
/coreclr/src/pal/inc/rt/poppack.h:37:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack()
        ^
In file included from /coreclr/src/pal/src/map/map.cpp:40:
/coreclr/src/pal/inc/rt/ntimage.h:571:10: error: the current #pragma pack aligment value is modified in the included
      file [-Werror,-Wpragma-pack]
#include "pshpack2.h"                       // Symbols, relocs, and linenumbers are 2 byte packed
         ^
/coreclr/src/pal/inc/rt/pshpack2.h:35:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack(2)
        ^
In file included from /coreclr/src/pal/src/map/map.cpp:40:
/coreclr/src/pal/inc/rt/ntimage.h:773:10: error: the current #pragma pack aligment value is modified in the included
      file [-Werror,-Wpragma-pack]
#include "poppack.h"
         ^
/coreclr/src/pal/inc/rt/poppack.h:37:9: note: previous '#pragma pack' directive that modifies alignment is here
#pragma pack()
        ^
```

**Comment by kasper3** [comment]
@janvorli, 6.0 was released today. New packages were made available (http://apt.llvm.org/).

They tend to introduce more analysis rules in every next version. Now 6.0 is stable and 7.0 is in development. Maybe it is worth adding a CI plan that runs once a day with nightly version of llvm in order to keep build future proof and reduced chance of breakage when a new version is officially released.

**Comment by janvorli** [comment]
This warning is generated for pattern that we use intentionally (we set / restore the packing by including pshpack/2/4/.h and poppack.h files that contain appropriate `#pragma pack` form based on the target compiler). The reason is that you cannot define a macro that would expand to `#pragma`. So we'll need to disable that warning. Ideally inside the above mentioned headers (disable before pushing the packing, enabling after popping it), but I guess it may not work since clang may then complain after the include of the poppack.h that the packing has changed since the warning would be enabled again.

**Comment by janvorli** [comment]
> Maybe it is worth adding a CI plan that runs once a day with nightly version of llvm in order to keep build future proof and reduced chance of breakage when a new version is officially released.

I am worried that it would generate too much churn - the compiler during its development phase may have issues that would break our build. And if we had such a daily run, someone would need to check each time a failure happens and try to figure out whether it is a problem of the compiler or of our code.
It seems better to keep monitoring the stable releases of clang and each time it is released, go and fix all the issues. That would consume much less time.

**Comment by jkotas** [comment]
I see we are spending a non-trivial amount of energy of fixing `-Werror` warnings. How much value are we getting back - how many real bugs do these warnings catch?

Should we rather be on http://blog.schmorp.de/2016-02-27-tidbits-for-the-love-of-god-dont-use-werror.html plan ?

**Comment by jkotas** [comment]
Maybe have the `-Werror` warnings enabled in CI only where we are building with fixed version of the compiler; but have them disabled by default when folks out there are building with random version of the compiler?

**Comment by janvorli** [comment]
Although I don't see the amount of energy spent of fixing the issues that large (basically max an hour each time a new compiler is released, at least so far), the plan to have -Werror enabled only for known versions of clang sounds good to me.

**Comment by janvorli** [comment]
Btw, I already have a fix for the clang 6.0 build ready.
